grpc/creds/creds.go

```diff
@@ -51,25 +54,148 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
+
+// NewServerTransport returns a new initialized grpc/credentials.TransportCredentials for server usage
+func NewServerTransport(serverConfig *tls.Config, acceptedSANs map[string]struct{}) credentials.TransportCredentials {
+	return &serverTransportCredentials{serverConfig, acceptedSANs}
+}
+
+// validateClient checks a peer's client certificate's SAN entries against
+// a list of accepted SANs. If the client certificate does not have a SAN on the
+// list it is rejected.
+//
+// Note 1: This function *only* verifies the SAN entries! Callers are expected to
+// have provided the `tls.ConnectionState` from returned from a validate (e.g.
+// non-error producing) `conn.Handshake()`.
+//
+// Note 2: We do *not* consider the client certificate subject common name. The
+// CN field is deprecated and should be present as a DNS SAN!
+func (tc *serverTransportCredentials) validateClient(peerState tls.ConnectionState) error {
+	// If there's no list of accepted SANs, all clients are OK
+	if tc.acceptedSANs == nil {
+		return nil
+	}
+
+	// If `conn.Handshake()` is called before `validateClient` this should not
+	// occur. We return an error in this event primarily for unit tests that may
+	// call `validateClient` with manufactured & artificial connection states.
+	if len(peerState.PeerCertificates) < 1 {
+		return fmt.Errorf(
+			"boulder/grpc/creds: validateClient given state with empty PeerCertificates")
+	}
+
+	// Since we call `conn.Handshake()` before `validateClient` and ensure
+	// a non-error response we don't need to validate anything except the presence
+	// of an accepted SAN in the leaf entry of `PeerCertificates`. The tls
+	// package's `serverHandshake` and in particular, `processCertsFromClient`
+	// will address everything else as an error returned from `Handshake()`.
+	var valid bool
+	leaf := peerState.PeerCertificates[0]
+
+	// First check the DNS subject alternate names against the accepted list
+	for _, dnsName := range leaf.DNSNames {
+		if _, ok := tc.acceptedSANs[dnsName]; ok {
+			valid = true
+		}
+	}
+	// Next check the IP address subject alternate names against the accepted list
+	for _, ip := range leaf.IPAddresses {
+		if _, ok := tc.acceptedSANs[ip.String()]; ok {
+			valid = true
+		}
+	}
+
+	// If none of the DNS or IP SANs on the leaf certificate matched the
+	// accepted list, the client isn't valid and we error
+	if !valid {
+		return fmt.Errorf(
+			"boulder/grpc/creds: peer's client certificate SAN entries did not match " +
+				"any entries on accepted SAN list.")
+	}
+
+	// Otherwise, the peer is valid!
+	return nil
+}
+
+// ServerHandshake does the authentication handshake for servers. It returns
+// the authenticated connection and the corresponding auth information about
+// the connection.
+func (tc *serverTransportCredentials) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	// If the `serverConfig` is nil we can't do anything further and must error
+	if tc.serverConfig == nil {
```

Comment:
This check should be in the `New...()` function instead. Then we can have an internal consistency guarantee that serverConfig is never nil, because the field is unexported.